Source/Dafny/Resolver/RunAllTestsMainMethod.cs

```diff
@@ -0,0 +1,225 @@
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using Microsoft.Boogie;
+
+namespace Microsoft.Dafny; 
+
+public class RunAllTestsMainMethod : IRewriter {
+
+  public RunAllTestsMainMethod(ErrorReporter reporter) : base(reporter) {
+  }
+
+  /// <summary>
+  /// Verifies that there isn't already a main method, and then creates
+  /// one with no body (to be filled in by PostResolve()).
+  ///
+  /// It might be possible to move this to PostResolve() if we created a resolved
+  /// Method instead, but for now this is a bit easier. It also allows us to produce
+  /// errors much earlier in the pipeline.
+  /// </summary>
+  internal override void PreResolve(Program program) {
+    Method mainMethod;
+    var hasMain = Compilers.SinglePassCompiler.HasMain(program, out mainMethod);
+    if (hasMain) {
+      Reporter.Error(MessageSource.Rewriter, mainMethod.tok, "Cannot use /runAllTests on a program with a main method");
```

Comment:
Would it make sense to replace the existing main method instead?

Reply:
I think it's generally better to catch cases of more than one main method by default, and I'd consider this an implicit case of that. I can imagine someone accidentally including source that includes an application main method, and this error would be an earlier signal their build is buggy. I've even accidentally included a dependency on a testing framework's CLI runner when I meant to depend on the package with the testing annotations, and the equivalent of this error saved me quite a bit of wasted time debugging. :)

Reply:
Makes sense, but… how do I run all tests for an application that *has* a main method?

Reply:
At least one approach is separating the source code into `src` and `test`, where the `test` code will reference most of the `src` content but not pull in the main method. The report generator does that for example: https://github.com/dafny-lang/dafny-reportgenerator. You could add `/runAllTests:1` to only the test project file here: https://github.com/dafny-lang/dafny-reportgenerator/blob/main/test/dafny-reportgenerator-test.csproj